nabs/preprocessors.py

```diff
@@ -301,4 +302,70 @@ def daq_during_plan():
     return (yield from daq_during_plan())
 
 
+def step_size_decorator(plan):
+    """
+    Grab the last argument (number of steps), and intepret as
+    - step size if float
+    - number of steps if integer
+
+    Only works on step scans in one dimension.
+
+    Parameters
+    ----------
+    plan : plan
+        A bluesky plan that yields bluesky Msg objects.  Must be
+        a scan in one dimension, with the last argument being
+        the number of scan points / step size.
+
+    Returns
+    -------
+    step_size_plan : plan
+        The same plan as before, but modified appropriately to
+        differentiate between step size and number of steps.
+        This will be a callable generator function.
+    """
+
+    @wraps(plan)
+    def inner(*args, **kwargs):
+        if 'num' in kwargs:
+            # Currently unneeded, since daq_ascan and daq_dscan
+            # do not support num kwarg
+            n = kwargs.pop('num')
+        else:
+            # assumes (det_list, motor, start, stop, num)
+            n = args[-1]
+
+        if not isinstance(n, (int, float)):
+            raise TypeError("Step size / number of steps is "
+                            "neither float nor integer")
+
+        if type(n) is int:
+            # interpret as number of steps (default)
+            yield from plan(*args, **kwargs)
+        elif type(n) is float:
+            # interpret as step size
+            start, stop = args[2], args[3]
+            # correct step size sign
+            n = np.sign(stop - start) * np.abs(n)
+            if np.abs(n) > np.abs(stop-start):
+                raise ValueError(f"Step size provided {n} greater "
+                                 "than the range provided "
+                                 f"{np.abs(stop-start)}")
+            step_list = list(np.arange(start, stop, n))
+            if len(step_list) == 0:
+                raise ValueError("Number of steps is 0 with the "
+                                 "provided range and step size.")
+
+            # new endpoint needed, for cases where
+            # (range % step_size) != 0 or to include endpoint
+            if np.isclose(step_list[-1] + n, stop):
+                step_list.append(step_list[-1] + n)
+            n_steps = len(step_list)
+
+            yield from plan(*args[:3], step_list[-1], n_steps,
+                            **kwargs)
```

Comment:
There's no `else:` clause here for contrived errors like `motor, 1, 10, "five"`

Reply:
I think I catch anything that's not an int or float early.

Reply:
My mistake - nice 😁